### DISCUSSION TITLE
Excluding hidden windows with `--match`

### DISCUSSION THREAD

**Discussion by crazygolem** [kovidgoyal/kitty#7886] [Q&A] [answered]
I am trying to match all visible windows in the active tab with `send-text` and ultimately `broadcast` (the former is easier to test with so I'll explain with it).

After looking at the documentation of both kittens, and also the discussion #5834, I think this window matcher should do the trick:

    state:parent_focused and not state:overlay_parent and not state:self

* `state:parent_focused` correctly selects only windows in the current tab
* `state:overlay_parent` and `state:self`, to my surprise, seem to match nothing (so the expression above matches everything)

For example, running the following command:

    kitten @send-text --match 'state:parent_focused and not state:overlay_parent and not state:self' 'date\n'

executes the `date` command in all windows of the current tab, including the window where the command was run, and another window that is hidden under an overlay.

[kitty-match.webm](https://github.com/user-attachments/assets/7d0901ef-caa0-43de-98bc-0ff68a712026)

Running it as a key binding:

    map f3 launch --type overlay --allow-remote-control kitty @ send-text --match 'state:parent_focused and not state:overlay_parent and not state:self' 'date\n'

still correctly targets windows in the current tab only, but also executes the command in the window where <kbd>F3</kbd> was pressed, and

    map f3 launch --type overlay --allow-remote-control kitty @ send-text --match 'state:overlay_parent or state:self' 'date\n'

outputs nothing.

I think I don't get what `state:overlay_parent` and `state:self` are supposed to match, can anyone clear this up for me?
And if you have any hint for achieving what I'm trying to do that would be swell =)

Thanks for your help :pray:

**Comment by kovidgoyal** [comment]
overlay_parent IIRC matches a window if it is under the self window
when the self window is an overlay. So not overlay_parent will match all
windows except the window under self if there is one.

not state:self should be working however. Try that in isolation without
overlay_parent.

I dont think there is a way to match visible windows excluding ones that
are below in an overlay stack. Would likely need a new state variable
for that, or you can always do the matching yourself by using kitten @
ls and selecting the windows based on whatever criteria you need.

**Reply by crazygolem** [reply]
Thanks a lot for your answer.

> overlay_parent IIRC matches a window if it is under the self window when the self window is an overlay. So not overlay_parent will match all windows except the window under self if there is one. not state:self should be working however. Try that in isolation without overlay_parent. I dont think there is a way to match visible windows excluding ones that are below in an overlay stack. Would likely need a new state variable for that, or you can always do the matching yourself by using kitten @ ls and selecting the windows based on whatever criteria you need.

Ah got it, I think I skipped over some words from the documentation :sweat_smile: 

> not state:self should be working however. Try that in isolation without overlay_parent.

Somehow it doesn't for me :(

1. I launch a kitty with remote-control enabled: `kitty --config NONE -o remote_control=yes`
2. I open a split: <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>enter</kbd>
3. Call `send-text` matching `not state:self`: all windows receive the text (in the screenshot: execute the `date` command)
4. Call `send-text` matching `state:self`: no window receive the text, including the one running the kitten.

![image](https://github.com/user-attachments/assets/99d5e0a4-8ddb-4e57-a8ce-f289c1c381b2)

I tried with a build from main with the same results, then I tried it in a freshly installed VM (with a version from the official repo) also with the same results. Could it be a GNOME+Wayland issue?
Would you like me to open a bug ticket to investigate further?

> I dont think there is a way to match visible windows excluding ones that are below in an overlay stack. Would likely need a new state variable for that, or you can always do the matching yourself by using kitten @ ls and selecting the windows based on whatever criteria you need.

Thanks for the hint.
It looks like the `groups` element under the tabs is what I'm looking for, where the visible windows would correspond to the last item of its `windows` element. I think I can work with that :+1: 

By the way it looks that the `--self` option to `kitten @ ls` also doesn't work for me, probably related to the match issue >.<

**Comment by kovidgoyal** [comment]
https://github.com/kovidgoyal/kitty/commit/958ad0d8b402698780808e45e075a28d9d15cc88

**Reply by crazygolem** [reply]
Fixes `--self` and `-m state:self`, thanks!

**Comment by crazygolem** [comment, marked as answer]
For posterity, the following `jq` query returns all the visible windows as a match expression:

```sh
kitten @ ls \
| jq -r '
  [ .[] | select(.is_focused)
  | .tabs[] | select(.is_focused)
  | .groups[].windows[-1]
  | "id:" + (. | tostring)
  ] | join(" or ")
  '
  ```

The result can then be incorporated into the `--match` option for `send-text`:

    "state:parent_focused and not state:self and ($tops)"

or for `broadcast` (the self window is automatically excluded by the kitten, and the previously visible window must now be additionally excluded if launched in an overlay):

    "state:parent_focused and not state:overlay_parent and ($tops)"
